internal/pkg/upload/upload.go

```diff
@@ -0,0 +1,288 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upload
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/elastic/fleet-server/v7/internal/pkg/bulk"
+	"github.com/elastic/go-elasticsearch/v8"
+	"github.com/gofrs/uuid"
+)
+
+const (
+	// specification-designated maximum
+	MaxChunkSize = 4194304 // 4 MiB
+)
+
+var (
+	ErrInvalidUploadID  = errors.New("active upload not found with this ID, it may be expired")
+	ErrFileSizeTooLarge = errors.New("this file exceeds the maximum allowed file size")
+	ErrMissingChunks    = errors.New("file data incomplete, not all chunks were uploaded")
+	ErrHashMismatch     = errors.New("hash does not match")
+	ErrUploadExpired    = errors.New("upload has expired")
+	ErrUploadStopped    = errors.New("upload has stopped")
+	ErrInvalidChunkNum  = errors.New("invalid chunk number")
+)
+
+// the only valid values of upload status according to storage spec
+type Status string
+
+const (
+	StatusAwaiting Status = "AWAITING_UPLOAD"
+	StatusProgress Status = "UPLOADING"
+	StatusDone     Status = "READY"
+	StatusFail     Status = "UPLOAD_ERROR"
+	StatusDel      Status = "DELETED"
+)
+
+type Info struct {
+	ID        string // upload operation identifier. Used to identify the upload process
+	DocID     string // document ID of the uploaded file and chunks
+	Source    string // which integration is performing the upload
+	AgentID   string
+	ActionID  string
+	ChunkSize int64
+	Total     int64
+	Count     int
+	Start     time.Time
+	Status    Status
+}
+
+// convenience functions for computing current "Status" based on the fields
+func (i Info) Expired(timeout time.Duration) bool { return time.Now().After(i.Start.Add(timeout)) }
+func (i Info) StatusCanUpload() bool { // returns true if more chunks can be uploaded. False if the upload process has completed (with or without error)
+	return !(i.Status == StatusFail || i.Status == StatusDone || i.Status == StatusDel)
+}
+
+type FileData struct {
+	Size      int64  `json:"size"`
+	ChunkSize int64  `json:"ChunkSize"`
+	Status    string `json:"Status"`
```

Comment:
this is deviating from the usual json naming convention where the rest of the fields user lower case letter and underscore "_" delimiter